File: archive/php/quote_to_image.php
<?php

// this script turns quotes from books into images for use in a Kindle clock.
// Jaap Meijers, 2018

require __DIR__ . '/vendor/autoload.php';

use Symfony\Component\Yaml\Yaml;

if (realpath(__FILE__) === realpath($_SERVER['SCRIPT_FILENAME'] ?? '')) {
    error_reporting(E_ALL);
    ini_set('display_errors', 1);
    ini_set('max_execution_time', 3000);

    if (!is_dir('images/metadata') && !mkdir('images/metadata', 0777, true)) {
        throw new \RuntimeException(sprintf('Directory "%s" was not created', 'images/metadata'));
    }

    $imagenumber = 0;
    $previoustime = 0;

    putenv('GDFONTPATH=' . realpath('.'));

    InitializeFonts();

    // Check for --all flag (include NSFW quotes); strip it before device parsing
    $includeNsfw = in_array('--all', $argv);
    $filteredArgv = array_values(array_filter($argv, fn($a) => $a !== '--all'));
    setDevice($filteredArgv);

    // get the quotes (including title and author) from the YAML file,
    // and create unique images for them, one without and one with title and author
    $entries = Yaml::parseFile('litclock.yaml');
    foreach ($entries as $entry) {
        // Skip NSFW quotes unless --all flag is set.
        // Entries without an 'sfw' field (original entries) are treated as SFW.
        if (!$includeNsfw && isset($entry['sfw']) && $entry['sfw'] === 'no') {
            continue;
        }

        $time       = $entry['time'];
        $timestring = trim($entry['time_name']);
        $quote      = trim(preg_replace('/\s+/', ' ', $entry['quote']));
        $title      = trim($entry['source']);
        $author     = trim($entry['author']);

        TurnQuoteIntoImage($time, $quote, $timestring, $title, $author);
    }
}

function setDevice($argv) {
    global $deviceWidth;
    global $deviceHeight;

    // default to Kindle size
    $deviceWidth = 600;
    $deviceHeight = 800;

    if (!empty($argv[1])) {
        $device = strtoupper($argv[1]);

        if ($device == "PAPERWHITE") {
            // Kindle Paperwhite 1–3 (older)
            $deviceWidth = 758;
            $deviceHeight = 1024;
        } elseif ($device == "PAPERWHITE5") {
            // Kindle Paperwhite 11th gen (2021), 6.8"
            $deviceWidth = 1236;
            $deviceHeight = 1648;
        } elseif ($device == "KINDLE" || $device == "BASIC") {
            // Kindle 11th gen (2022), 6"
            $deviceWidth = 1072;
            $deviceHeight = 1448;
        } elseif ($device == "OASIS") {
            // Kindle Oasis (10th gen), 7"
            $deviceWidth = 1264;
            $deviceHeight = 1680;
        } elseif ($device == "SCRIBE") {
            // Kindle Scribe, 10.2"
            $deviceWidth = 1860;
            $deviceHeight = 2480;
        } elseif ($device == "CLARA") {
            // Kobo Clara 2E / Clara BW, 6"
            $deviceWidth = 1072;
            $deviceHeight = 1448;
        } elseif ($device == "LIBRA") {
            // Kobo Libra 2 / Libra Colour, 7"
            $deviceWidth = 1264;
            $deviceHeight = 1680;
        } elseif ($device == "SAGE") {
            // Kobo Sage, 8"
            $deviceWidth = 1440;
            $deviceHeight = 1920;
        } elseif ($device == "ELIPSA") {
            // Kobo Elipsa 2E, 10.3"
            $deviceWidth = 1404;
            $deviceHeight = 1872;
        } elseif ($device == "REMARKABLE") {
            // reMarkable 2, 10.3"
            $deviceWidth = 1404;
            $deviceHeight = 1872;
        } elseif ($device == "REMARKABLEPRO") {
            // reMarkable Paper Pro, 11.8"
            $deviceWidth = 1620;
            $deviceHeight = 2160;
        } elseif ($device == "INKYPHAT") {
            // Inky pHAT (portrait)
            $deviceWidth = 104;
            $deviceHeight = 212;
        } elseif ($device == "INKYPHAT_L") {
            // Inky pHAT (landscape)
            $deviceWidth = 212;
            $deviceHeight = 104;
        } elseif ($device == "INKYWHAT") {
            // Inky wHAT (portrait)
            $deviceWidth = 300;
            $deviceHeight = 400;
        } elseif ($device == "INKYWHAT_L") {
            // Inky wHAT (landscape)
            $deviceWidth = 400;
            $deviceHeight = 300;
        } elseif ($device == "INKYIMPRESSION") {
            // Inky Impression (portrait)
            $deviceWidth = 448;
            $deviceHeight = 600;
        } elseif ($device == "INKYIMPRESSION_L") {
            // Inky Impression (landscape)
            $deviceWidth = 600;
            $deviceHeight = 448;
        } elseif ($device == "WAVESHARE75") {
            // Waveshare 7.5" (portrait)
            $deviceWidth = 480;
            $deviceHeight = 800;
        } elseif ($device == "WAVESHARE75_L") {
            // Waveshare 7.5" (landscape)
            $deviceWidth = 800;
            $deviceHeight = 480;
        } elseif ($device == "IT8951") {
            // IT8951 10.3" e-ink panel (portrait)
            $deviceWidth = 1404;
            $deviceHeight = 1872;
        } elseif ($device == "IT8951_L") {
            // IT8951 10.3" e-ink panel (landscape)
            $deviceWidth = 1872;
            $deviceHeight = 1404;
        } elseif ($device == "CUSTOM") {
            $deviceWidth = $argv[2];
            $deviceHeight = $argv[3];
        }
        // unrecognised device falls through to defaults set above
    }
}

function resolveFont($ttf, $otf) {
    if (file_exists($ttf)) {
        return $ttf;
    } elseif (file_exists($otf)) {
        return $otf;
    }
    print "ERROR: Unable to find font file: " . $ttf . " or " . $otf . PHP_EOL;
    return null;
}

function InitializeFonts()
{
    global $font_path;
    global $font_path_bold;
    global $creditFont;

    $font_path      = resolveFont('LinLibertine_RZ.ttf',  'LinLibertine_RZ.otf');
    $font_path_bold = resolveFont('LinLibertine_RB.ttf',  'LinLibertine_RB.otf');
    $creditFont     = resolveFont('LinLibertine_RZI.ttf', 'LinLibertine_RZI.otf');

    if ($font_path === null || $font_path_bold === null || $creditFont === null) {
        print "ERROR loading fonts!" . PHP_EOL;
        print "Please download and install the fonts from here: https://sourceforge.net/projects/linuxlibertine/ into the current directory." . PHP_EOL;
        exit;
    }
}


function TurnQuoteIntoImage($time, $quote, $timestring, $title, $author)
{
    global $font_path;
    global $font_path_bold;
    global $creditFont;
    global $deviceWidth;
    global $deviceHeight;

    $width  = $deviceWidth;
    $height = $deviceHeight;
    $margin = 26;

    // find the position and length of the timestring within the quote
    $timestringStarts     = count(explode(' ', stristr($quote, $timestring, true))) - 1;
    $timestring_wordcount = count(explode(' ', $timestring)) - 1;
    $quote_array          = explode(' ', $quote);

    $time      = substr($time, 0, 2) . substr($time, 3, 2);
    $font_size = 18;

    // serial number for when there is more than one quote for a certain minute
    global $imagenumber;
    global $previoustime;
    if ($time == $previoustime) {
        $imagenumber++;
    } else {
        $imagenumber = 0;
    }
    $previoustime = $time;

    $quotePath   = 'images/quote_' . $time . '_' . $imagenumber . '.png';
    $creditsPath = 'images/metadata/quote_' . $time . '_' . $imagenumber . '_credits.png';

    // Does the image already exist? No point in creating it again.
    if (file_exists($quotePath)) {
        return;
    }

    printf('Making a new image for ' . $time . PHP_EOL);
    if (PHP_SAPI !== 'cli') echo '<br />';

    ///// QUOTE /////
    // find the font size (recursively) for an optimal fit and create the image
    list($png_image) = fitText($quote_array, $width, $height, $font_size, $timestringStarts, $timestring_wordcount, $margin);

    print 'Image for ' . $time . '_' . $imagenumber . PHP_EOL;
    if (PHP_SAPI !== 'cli') echo '<br /><br />';

    imagepng($png_image, $quotePath);

    ///// METADATA /////
    // create another version, with title and author in the image
    $black   = imagecolorallocate($png_image, 0, 0, 0);
    $dash    = '—';
    $credits = $title . ', ' . $author;
    $creditFont_size = 18;

    list($metawidth, , $metaleft) = measureSizeOfTextbox($creditFont_size, $creditFont, $dash . $credits);

    if ($metawidth > 500) {
        $creditsArray = explode(' ', $credits);
        $newCredits   = array();
        $i = 1;

        while (true) {
            $tmp0 = implode(' ', array_slice($creditsArray, 0, count($creditsArray) - $i));
            $tmp1 = implode(' ', array_slice($creditsArray, 0 - $i));

            // once the second line is (almost) longer than the first, stop
            if (strlen($tmp1) + 5 > strlen($tmp0)) {
                break;
            }
            $newCredits[0] = $tmp0;
            $newCredits[1] = $tmp1;
            $i++;
        }

        list($textWidth1, $textheight1) = measureSizeOfTextbox($creditFont_size, $creditFont, $dash . $newCredits[0]);
        list($textWidth2)               = measureSizeOfTextbox($creditFont_size, $creditFont, $newCredits[1]);

        $metadataX1 = $width - ($textWidth1 + $margin);
        $metadataX2 = $width - ($textWidth2 + $margin);
        $metadataY  = $height - $margin;

        imagettftext($png_image, $creditFont_size, 0, $metadataX1, $metadataY - (int)($textheight1 * 1.1), $black, $creditFont, $dash . $newCredits[0]);
        imagettftext($png_image, $creditFont_size, 0, $metadataX2, $metadataY, $black, $creditFont, $newCredits[1]);
    } else {
        $metadataX = ($width - $metaleft) - $margin;
        $metadataY = $height - $margin;
        imagettftext($png_image, $creditFont_size, 0, $metadataX, $metadataY, $black, $creditFont, $dash . $credits);
    }

    imagepng($png_image, $creditsPath);
    imagedestroy($png_image);
}


// Measures text layout without creating an image. Returns paragraph height, or false if a word is too wide.
function measureLayout($quote_array, $width, $font_size, $timestringStarts, $timestring_wordcount, $margin)
{
    global $font_path_bold;
    global $font_path;

    $position = [$margin, $margin + $font_size];

    foreach ($quote_array as $key => $word) {
        $font = ($key >= $timestringStarts && $key <= $timestringStarts + $timestring_wordcount)
            ? $font_path_bold
            : $font_path;

        list($textwidth) = measureSizeOfTextbox($font_size, $font, $word . ' ');

        if ($textwidth > ($width - $margin)) {
            return false; // single word too wide at this size
        }

        if (($position[0] + $textwidth) >= ($width - $margin)) {
            $position[0] = $margin;
            $position[1] += (int)($font_size * 1.618); // golden ratio line height
        }

        $position[0] += $textwidth;
    }

    return $position[1];
}

// Creates the image at a specific font size.
function renderImage($quote_array, $width, $height, $font_size, $timestringStarts, $timestring_wordcount, $margin)
{
    global $font_path_bold;
    global $font_path;

    $png_image = imagecreate($width, $height)
        or die('Cannot Initialize new GD image stream');
    imagecolorallocate($png_image, 255, 255, 255); // background (first allocation sets fill color)

    $grey  = imagecolorallocate($png_image, 125, 125, 125);
    $black = imagecolorallocate($png_image, 0, 0, 0);

    $position = [$margin, $margin + $font_size];

    foreach ($quote_array as $key => $word) {
        // use bold + black for the timestring words, grey for the rest
        if ($key >= $timestringStarts && $key <= $timestringStarts + $timestring_wordcount) {
            $font      = $font_path_bold;
            $textcolor = $black;
        } else {
            $font      = $font_path;
            $textcolor = $grey;
        }

        list($textwidth) = measureSizeOfTextbox($font_size, $font, $word . ' ');

        if (($position[0] + $textwidth) >= ($width - $margin)) {
            $position[0] = $margin;
            $position[1] += (int)($font_size * 1.618); // golden ratio line height
        }

        imagettftext($png_image, $font_size, 0, $position[0], $position[1], $textcolor, $font, $word);
        $position[0] += $textwidth;
    }

    return [$png_image, $position[1], $font_size];
}

function fitText($quote_array, $width, $height, $font_size, $timestringStarts, $timestring_wordcount, $margin)
{
    // Find the largest font size that fits using measurements only (no image creation).
    $best_size = null;

    for ($size = $font_size; ; $size++) {
        $paragraphHeight = measureLayout($quote_array, $width, $size, $timestringStarts, $timestring_wordcount, $margin);
        if ($paragraphHeight === false || $paragraphHeight >= $height - 100) {
            break;
        }
        $best_size = $size;
    }

    if ($best_size === null) {
        return false;
    }

    return renderImage($quote_array, $width, $height, $best_size, $timestringStarts, $timestring_wordcount, $margin);
}

function measureSizeOfTextbox($font_size, $font_path, $text)
{
    $box = imagettfbbox($font_size, 0, $font_path, $text);

    $min_x = min($box[0], $box[2], $box[4], $box[6]);
    $max_x = max($box[0], $box[2], $box[4], $box[6]);
    $min_y = min($box[1], $box[3], $box[5], $box[7]);
    $max_y = max($box[1], $box[3], $box[5], $box[7]);

    $width  = ($max_x - $min_x);
    $height = ($max_y - $min_y);
    $left   = abs($min_x) + $width;
    $top    = abs($min_y) + $height;

    return array($width, $height, $left, $top);
}

File: quote_to_image.py
#!/usr/bin/env python3
"""Generate literary clock images from litclock.yaml."""

import argparse
import os
import sys
from functools import partial
from multiprocessing import Pool, cpu_count

import yaml
from PIL import Image, ImageDraw, ImageFont

DEVICE_PRESETS = {
    'kindle':           (600,  800),
    'basic':            (1072, 1448),
    'paperwhite':       (758,  1024),
    'paperwhite5':      (1236, 1648),
    'oasis':            (1264, 1680),
    'scribe':           (1860, 2480),
    'clara':            (1072, 1448),
    'libra':            (1264, 1680),
    'sage':             (1440, 1920),
    'elipsa':           (1404, 1872),
    'remarkable':       (1404, 1872),
    'remarkablepro':    (1620, 2160),
    'inkyphat':         (104,  212),
    'inkyphat_l':       (212,  104),
    'inkywhat':         (300,  400),
    'inkywhat_l':       (400,  300),
    'inkyimpression':   (448,  600),
    'inkyimpression_l': (600,  448),
    'waveshare75':      (480,  800),
    'waveshare75_l':    (800,  480),
    'it8951':           (1404, 1872),
    'it8951_l':         (1872, 1404),
}


def resolve_font(ttf, otf):
    """Return ttf path if it exists, otf path if not, or None if neither."""
    if os.path.exists(ttf):
        return ttf
    if os.path.exists(otf):
        return otf
    print(f'ERROR: Unable to find font file: {ttf} or {otf}', file=sys.stderr)
    return None


def measure_layout(words, width, font_size, time_start, time_count, margin,
                   regular_path, bold_path):
    """
    Simulate word layout at font_size without creating an image.
    Returns the final y position (paragraph height), or None if a single word is too wide.
    """
    regular = ImageFont.truetype(regular_path, font_size)
    bold    = ImageFont.truetype(bold_path,    font_size)

    x = margin
    y = margin + font_size

    for i, word in enumerate(words):
        font = bold if time_start <= i <= time_start + time_count else regular
        w = int(font.getlength(word + ' '))

        if w > width - margin:
            return None  # single word too wide at this size

        if x + w >= width - margin:
            x  = margin
            y += int(font_size * 1.618)  # golden ratio line height

        x += w

    return y


def render_image(words, width, height, font_size, time_start, time_count, margin,
                 regular_path, bold_path):
    """
    Render words onto a greyscale PIL Image at the given font_size.
    Time-string words: black + bold. All others: grey + regular.
    """
    regular = ImageFont.truetype(regular_path, font_size)
    bold    = ImageFont.truetype(bold_path,    font_size)

    img  = Image.new('L', (width, height), 255)
    draw = ImageDraw.Draw(img)

    x = margin
    y = margin + font_size

    for i, word in enumerate(words):
        if time_start <= i <= time_start + time_count:
            font  = bold
            color = 0    # black
        else:
            font  = regular
            color = 125  # grey

        w = int(font.getlength(word + ' '))

        if x + w >= width - margin:
            x  = margin
            y += int(font_size * 1.618)

        draw.text((x, y), word, font=font, fill=color, anchor='ls')
        x += w

    return img


def fit_text(words, width, height, font_size, time_start, time_count, margin,
             regular_path, bold_path):
    """
    Find the largest font size where text fits within height - 100.
    Returns a rendered PIL Image, or None if the minimum size doesn't fit.
    """
    best = None
    size = font_size

    while True:
        h = measure_layout(words, width, size, time_start, time_count, margin,
                           regular_path, bold_path)
        if h is None or h >= height - 100:
            break
        best = size
        size += 1

    if best is None:
        return None

    return render_image(words, width, height, best, time_start, time_count, margin,
                        regular_path, bold_path)


def _draw_credits(img, width, height, margin, credit_path, source, author):
    """Draw right-aligned credits onto img in-place."""
    draw        = ImageDraw.Draw(img)
    credit_font = ImageFont.truetype(credit_path, 18)
    em_dash     = '\u2014'
    credits     = f'{source}, {author}'
    full_text   = f'{em_dash}{credits}'
    text_width  = int(credit_font.getlength(full_text))

    if text_width > 500:
        # split credits (without dash) into two balanced lines
        words_c = credits.split()
        line1, line2 = credits, ''
        for i in range(1, len(words_c)):
            l1 = ' '.join(words_c[:-i])
            l2 = ' '.join(words_c[-i:])
            if len(l2) + 5 > len(l1):
                break
            line1, line2 = l1, l2

        full_line1  = f'{em_dash}{line1}'
        w1          = int(credit_font.getlength(full_line1))
        w2          = int(credit_font.getlength(line2))
        bbox1       = credit_font.getbbox(full_line1)
        line_height = int((bbox1[3] - bbox1[1]) * 1.1)

        draw.text((width - w1 - margin, height - margin - line_height),
                  full_line1, font=credit_font, fill=0, anchor='ls')
        draw.text((width - w2 - margin, height - margin),
                  line2, font=credit_font, fill=0, anchor='ls')
    else:
        draw.text((width - text_width - margin, height - margin),
                  full_text, font=credit_font, fill=0, anchor='ls')


def render_entry(entry, width, height, font_paths):
    """
    Render quote and credits images for one entry.
    Top-level function (required for multiprocessing pickling).
    """
    regular_path, bold_path, credit_path = font_paths

    time_key = entry['time'].replace(':', '')
    idx      = entry['idx']

    quote_out   = os.path.join('images', f'quote_{time_key}_{idx}.png')
    credits_out = os.path.join('images', 'metadata', f'quote_{time_key}_{idx}_credits.png')

    if os.path.exists(quote_out) and os.path.exists(credits_out):
        return

    margin     = 26
    timestring = entry['time_name'].strip()
    quote      = ' '.join(entry['quote'].split())  # normalise whitespace
    words      = quote.split()

    # find word index where the time string begins (case-insensitive)
    lower_quote = quote.lower()
    lower_time  = timestring.lower()
    pos = lower_quote.find(lower_time)
    if pos == -1:
        time_start = 0
        time_count = 0
    else:
        # snap back to the nearest word boundary so a mid-word pos doesn't
        # cause before.split() to count the preceding partial word
        while pos > 0 and quote[pos - 1] != ' ':
            pos -= 1
        before     = quote[:pos]
        time_start = len(before.split())
        time_count = len(timestring.split()) - 1

    print(f'Making image for {time_key}_{idx}')

    img = fit_text(words, width, height, 18, time_start, time_count, margin,
                   regular_path, bold_path)
    if img is None:
        print(f'WARNING: could not fit text for {time_key}_{idx}, skipping',
              file=sys.stderr)
        return

    img.save(quote_out)

    _draw_credits(img, width, height, margin, credit_path,
                  entry['source'].strip(), entry['author'].strip())
    img.save(credits_out)


def main():
    parser = argparse.ArgumentParser(
        description='Generate literary clock images from litclock.yaml.')
    parser.add_argument(
        '--device', default='kindle',
        help=f'Device preset, case-insensitive (default: kindle). '
             f'Known: {", ".join(DEVICE_PRESETS)}')
    parser.add_argument('--width',   type=int, help='Override image width in pixels')
    parser.add_argument('--height',  type=int, help='Override image height in pixels')
    parser.add_argument('--all',     action='store_true', dest='include_nsfw',
                        help='Include NSFW quotes (omitted by default)')
    parser.add_argument('--workers', type=int, default=cpu_count(),
                        help='Parallel worker processes (default: CPU count)')
    args = parser.parse_args()

    device = args.device.lower()
    if device not in DEVICE_PRESETS:
        print(f"ERROR: Unknown device '{args.device}'. "
              f"Known: {', '.join(DEVICE_PRESETS)}", file=sys.stderr)
        sys.exit(1)

    width, height = DEVICE_PRESETS[device]
    if args.width is not None:
        width = args.width
    if args.height is not None:
        height = args.height

    regular_path = resolve_font('LinLibertine_RZ.ttf',  'LinLibertine_RZ.otf')
    bold_path    = resolve_font('LinLibertine_RB.ttf',  'LinLibertine_RB.otf')
    credit_path  = resolve_font('LinLibertine_RZI.ttf', 'LinLibertine_RZI.otf')

    if None in (regular_path, bold_path, credit_path):
        print('ERROR: Missing font files. Download LinLibertine from '
              'https://sourceforge.net/projects/linuxlibertine/', file=sys.stderr)
        sys.exit(1)

    os.makedirs(os.path.join('images', 'metadata'), exist_ok=True)

    with open('litclock.yaml', encoding='utf-8') as f:
        entries = yaml.safe_load(f)

    if not args.include_nsfw:
        entries = [e for e in entries if e.get('sfw', 'yes') != 'no']

    # assign indices before parallelising — order must be preserved
    time_counts: dict = {}
    tasks = []
    for entry in entries:
        t   = entry['time'].replace(':', '')
        idx = time_counts.get(t, 0)
        time_counts[t] = idx + 1
        tasks.append({**entry, 'idx': idx})

    font_paths = (regular_path, bold_path, credit_path)
    worker     = partial(render_entry, width=width, height=height, font_paths=font_paths)

    with Pool(args.workers) as pool:
        pool.map(worker, tasks)


if __name__ == '__main__':
    main()
